x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiStreamingProcessor.java

```diff
@@ -0,0 +1,203 @@
+ *              "object": "chat.completion",
+ *              "created": 1677652288,
+ *              "model": "gpt-3.5-turbo-0613",
+ *              "system_fingerprint": "fp_44709d6fcb",
+ *              "choices": [
+ *                  {
+ *                      "index": 2,
+ *                      "delta": {},
+ *                      "finish_reason": "stop"
+ *                  }
+ *              ]
+ *          }
+ *
+ *          [DONE]
+ *     </code>
+ * </pre>
+ */
+public class OpenAiStreamingProcessor extends DelegatingProcessor<Deque<ServerSentEvent>, ChunkedToXContent> {
+    private static final Logger log = LogManager.getLogger(OpenAiStreamingProcessor.class);
+    private static final String FAILED_TO_FIND_FIELD_TEMPLATE = "Failed to find required field [%s] in OpenAI chat completions response";
+    private static final String RESULT = "delta";
+
+    private static final String CHOICES_FIELD = "choices";
+    private static final String DELTA_FIELD = "delta";
+    private static final String CONTENT_FIELD = "content";
+    private static final String FINISH_REASON_FIELD = "finish_reason";
+    private static final String STOP_MESSAGE = "stop";
+    private static final String DONE_MESSAGE = "[done]";
+
+    @Override
+    protected void next(Deque<ServerSentEvent> item) {
+        var parserConfig = XContentParserConfiguration.EMPTY.withDeprecationHandler(LoggingDeprecationHandler.INSTANCE);
+
+        var results = new ArrayDeque<ChunkedToXContent>(item.size());
+        for (ServerSentEvent event : item) {
+            if (ServerSentEventField.DATA == event.name() && event.hasValue()) {
+                try {
+                    var delta = parse(parserConfig, event);
+                    delta.map(this::deltaChunk).ifPresent(results::offer);
+                } catch (Exception e) {
+                    log.warn("Failed to parse event from inference provider: {}", event);
+                    onError(new IOException("Failed to parse event from inference provider.", e));
+                    return;
+                }
+            }
+        }
+
+        if (results.isEmpty()) {
+            upstream().request(1);
+        } else {
+            downstream().onNext(completionChunk(results.iterator()));
+        }
+    }
+
+    private Optional<String> parse(XContentParserConfiguration parserConfig, ServerSentEvent event) throws IOException {
+        if (DONE_MESSAGE.equalsIgnoreCase(event.value())) {
+            return Optional.empty();
+        }
+
+        try (XContentParser jsonParser = XContentFactory.xContent(XContentType.JSON).createParser(parserConfig, event.value())) {
+            moveToFirstToken(jsonParser);
+
+            XContentParser.Token token = jsonParser.currentToken();
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, token, jsonParser);
+
+            // choices is an array, but since we don't send 'n' in the request then we only get one value in the result
```

Comment:
Supporting `n > 1` for choices would make the response structure very complicated. We just need to document that only the first choice is streamed